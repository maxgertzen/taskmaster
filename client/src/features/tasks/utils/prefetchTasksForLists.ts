import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
