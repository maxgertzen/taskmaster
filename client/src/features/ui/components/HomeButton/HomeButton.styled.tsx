import styled from '@emotion/styled';

export const StyledHomeButton = styled('div')(({ theme }) => ({
  display: 'flex',
  justifyContent: 'space-between',
  alignItems: 'center',
  gap: theme.spacing(2),
  cursor: 'pointer',
}));

export const TitleContainer = styled('div')(({ theme }) => ({
  display: 'flex',
  alignItems: 'center',
  gap: theme.spacing(1),
}));
