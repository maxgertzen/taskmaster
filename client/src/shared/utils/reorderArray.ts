export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
