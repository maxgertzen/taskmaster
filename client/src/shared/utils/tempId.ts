export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
