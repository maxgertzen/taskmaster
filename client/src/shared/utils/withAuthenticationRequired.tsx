import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
