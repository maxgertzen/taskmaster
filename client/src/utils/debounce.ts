export function debounce<T extends (...args: Parameters<T>) => ReturnType<T>>(
  func: T,
  delay: number
): ((...args: Parameters<T>) => void) & { cancel: () => void } {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  const debouncedFunction = (...args: Parameters<T>) => {
    if (timeoutId) clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };

  debouncedFunction.cancel = () => {
    if (timeoutId) clearTimeout(timeoutId);
  };

  return debouncedFunction;
}
