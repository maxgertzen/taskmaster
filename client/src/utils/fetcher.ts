export type FetcherConfig<K> = {
  url: string;
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  body?: K;
  timeout?: number;
  headers?: HeadersInit;
  urlSearchParams?: string;
};

export const fetcher = async <T, K = undefined>(
  config: FetcherConfig<K>
): Promise<T> => {
  try {
    const {
      method = 'GET',
      body = {},
      headers = {},
      timeout = 5000,
      urlSearchParams,
    } = config;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    const options: RequestInit = {
      method,
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
      ...(method !== 'GET' && { body: JSON.stringify(body) }),
      signal: controller.signal,
    };

    if (urlSearchParams) {
      config.url += `?${urlSearchParams}`;
    }

    const response = await fetch(config.url, options);
    clearTimeout(timeoutId);

    if (!response.ok) {
      throw new Error(`Error: ${response.status} ${response.statusText}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Failed to fetch data:', error);
    throw error;
  }
};
