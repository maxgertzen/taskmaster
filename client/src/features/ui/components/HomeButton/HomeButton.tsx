import { FC } from 'react';

import { SpriteIcon, Title } from '..';

import { StyledHomeButton, TitleContainer } from './HomeButton.styled';

interface HomeButtonProps {
  isCollapsed: boolean;
  onClick: () => void;
}

export const HomeButton: FC<HomeButtonProps> = ({ isCollapsed, onClick }) => {
  return (
    <StyledHomeButton data-testid='home-button' role='button' onClick={onClick}>
      <TitleContainer>
        <SpriteIcon name='home' />
        {!isCollapsed && <Title variant='h6'>Home</Title>}
      </TitleContainer>
    </StyledHomeButton>
  );
};
