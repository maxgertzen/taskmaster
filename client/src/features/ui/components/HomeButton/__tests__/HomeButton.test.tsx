import { act, fireEvent } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach } from 'vitest';

import { render, screen } from '@/tests/utils';

import { HomeButton } from '../HomeButton';

describe('HomeButton', () => {
  const mockOnClick = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe.each([
    ['expanded', false],
    ['collapsed', true],
  ])('when %s (isCollapsed: %s)', (_, isCollapsed) => {
    it.each([
      ['light', undefined],
      ['dark', 'dark'],
    ])('renders correctly in %s theme', (_, theme) => {
      const { container } = render(
        <HomeButton isCollapsed={isCollapsed} onClick={mockOnClick} />,
        { theme: theme as 'light' | 'dark' }
      );
      expect(container).toMatchSnapshot();
    });

    it('handles text visibility correctly', () => {
      render(<HomeButton isCollapsed={isCollapsed} onClick={mockOnClick} />);

      if (isCollapsed) {
        expect(screen.queryByText('Home')).not.toBeInTheDocument();
      } else {
        expect(screen.getByText('Home')).toBeInTheDocument();
      }
    });
  });

  describe('User Interactions', () => {
    it('calls onClick handler when clicked', async () => {
      render(<HomeButton isCollapsed={false} onClick={mockOnClick} />);

      await act(async () => {
        fireEvent.click(screen.getByTestId('home-button'));
      });

      expect(mockOnClick).toHaveBeenCalledTimes(1);
    });
  });

  describe('Icon Rendering', () => {
    it('always renders home icon regardless of collapsed state', () => {
      const { rerender } = render(
        <HomeButton isCollapsed={false} onClick={mockOnClick} />
      );

      expect(screen.getByTestId('home-icon')).toBeInTheDocument();

      rerender(<HomeButton isCollapsed={true} onClick={mockOnClick} />);

      expect(screen.getByTestId('home-icon')).toBeInTheDocument();
    });
  });
});
