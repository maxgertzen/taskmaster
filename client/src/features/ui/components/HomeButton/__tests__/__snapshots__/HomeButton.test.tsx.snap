// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomeButton > when collapsed (isCollapsed: true) > renders correctly in dark theme 1`] = `
<div>
  <div
    class="css-1jc2zkj"
    data-testid="home-button"
    role="button"
  >
    <div
      class="css-axw7ok"
    >
      <div
        class="css-4dcg29"
        data-testid="icon-home"
        role="button"
        size="3"
      >
        <svg
          data-testid="home-icon"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`HomeButton > when collapsed (isCollapsed: true) > renders correctly in light theme 1`] = `
<div>
  <div
    class="css-1jc2zkj"
    data-testid="home-button"
    role="button"
  >
    <div
      class="css-axw7ok"
    >
      <div
        class="css-4dcg29"
        data-testid="icon-home"
        role="button"
        size="3"
      >
        <svg
          data-testid="home-icon"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`HomeButton > when expanded (isCollapsed: false) > renders correctly in dark theme 1`] = `
<div>
  <div
    class="css-1jc2zkj"
    data-testid="home-button"
    role="button"
  >
    <div
      class="css-axw7ok"
    >
      <div
        class="css-4dcg29"
        data-testid="icon-home"
        role="button"
        size="3"
      >
        <svg
          data-testid="home-icon"
        />
      </div>
      <h6
        css="[object Object]"
      >
        Home
      </h6>
    </div>
  </div>
</div>
`;

exports[`HomeButton > when expanded (isCollapsed: false) > renders correctly in light theme 1`] = `
<div>
  <div
    class="css-1jc2zkj"
    data-testid="home-button"
    role="button"
  >
    <div
      class="css-axw7ok"
    >
      <div
        class="css-4dcg29"
        data-testid="icon-home"
        role="button"
        size="3"
      >
        <svg
          data-testid="home-icon"
        />
      </div>
      <h6
        css="[object Object]"
      >
        Home
      </h6>
    </div>
  </div>
</div>
`;
